test/boost/aws_error_injection_test.cc

```diff
@@ -0,0 +1,195 @@
+#include <seastar/core/fstream.hh>
+#include <seastar/core/units.hh>
+#include <seastar/http/httpd.hh>
+
+using namespace seastar;
+using namespace std::chrono_literals;
+struct server {
+    enum class failure_policy : uint8_t {
+        SUCCESS = 0,
+        RETRYABLE_FAILURE = 1,
+        NONRETRYABLE_FAILURE = 2,
+    };
+    class dummy_aws_request_handler : public httpd::handler_base {
+    public:
+        explicit dummy_aws_request_handler(failure_policy _test_failure_policy) : test_failure_policy(_test_failure_policy) {}
+        future<std::unique_ptr<http::reply>> handle(const sstring& path, std::unique_ptr<http::request> req, std::unique_ptr<http::reply> rep) override {
+            auto method = req->_method;
+            auto url_params = req->query_parameters;
+            testlog.debug("{}\t{}", req->_method, req->get_url());
+            rep->write_body("txt", build_response(*req));
+            if (method == "DELETE")
+                rep->set_status(http::reply::status_type::no_content);
+            else if (method == "PUT") {
+                rep->set_status(http::reply::status_type::ok);
+                if (url_params.contains("partNumber") && url_params.contains("uploadId")) {
+                    rep->add_header("ETag", "SomeTag_" + url_params.at("partNumber"));
+                }
+                rep->add_header("ETag", "SomeTag");
+            }
+            return make_ready_future<std::unique_ptr<http::reply>>(std::move(rep));
+        }
+
+    private:
+        sstring build_response(const http::request& req) {
+            sstring response;
+            if (req._method == "POST") {
+                if (req.query_parameters.contains("uploads")) {
+                    response = "<InitiateMultipartUploadResult><Bucket>bucket</Bucket><Key>key</Key><UploadId>UploadId</UploadId></"
+                               "InitiateMultipartUploadResult>";
+                } else if (req.query_parameters.contains("uploadId")) {
+                    switch (test_failure_policy) {
+                    case failure_policy::SUCCESS:
+                        return R"(<CompleteMultipartUploadResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
+                                     <Location>http://Example-Bucket.s3.Region.amazonaws.com/Example-Object</Location>
+                                     <Bucket>Example-Bucket</Bucket>
+                                     <Key>Example-Object</Key>
+                                     <ETag>"3858f62230ac3c915f300c664312c11f-9"</ETag>
+                                </CompleteMultipartUploadResult>)";
+                    case failure_policy::RETRYABLE_FAILURE:
+                        // should succeed on retry
+                        test_failure_policy = failure_policy::SUCCESS;
+                        return R"(<?xml version="1.0" encoding="UTF-8"?>
+
+                                 <Error>
+                                  <Code>InternalError</Code>
```

Comment:
It's not immediately obvious that InternalError implies that it will be retried. A static_assert on `{"InternalError", aws_error(aws_error_type::INTERNAL_FAILURE, true)},` this line from `aws_error_map` would help

Reply:
I dont see how static_assert can be used here, will add regular `assert`

Reply:
Done